src/backend/distributed/utils/tuplestore.c

```diff
@@ -0,0 +1,98 @@
+/*-------------------------------------------------------------------------
+ *
+ * tuplestore.c
+ *	  Utilities regarding calls to PG functions
+ *
+ * Copyright (c) 2012-2018, Citus Data, Inc.
+ *-------------------------------------------------------------------------
+ */
+
+
+#include "postgres.h"
+
+#include "distributed/tuplestore.h"
+#include "miscadmin.h"
+
+/*
+ * CheckTuplestoreReturn checks if a tuplestore can be returned in the callsite
+ * of the UDF.
+ */
+ReturnSetInfo *
+CheckTuplestoreReturn(FunctionCallInfo fcinfo, TupleDesc *tupdesc)
+{
+	/* IMPORTANT can this be undefined behaviour. Casting to ReturnSetInfo before
+	 * we checked the type with IsA?
+	 * In case the type actual type is more strictly aligned than ReturnSetInfo
+	 * it is it seems:
+	 * https://wiki.sei.cmu.edu/confluence/display/c/EXP36-C.+Do+not+cast+pointers+into+more+strictly+aligned+pointer+types */
+	ReturnSetInfo *returnSetInfo = (ReturnSetInfo *) fcinfo->resultinfo;
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (returnSetInfo == NULL || !IsA(returnSetInfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context that cannot " \
+						"accept a set")));
+	}
+	if (!(returnSetInfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not " \
+						"allowed in this context")));
+	}
+	switch (get_call_result_type(fcinfo, NULL, tupdesc))
+	{
+		case TYPEFUNC_COMPOSITE:
+		{
+			/* success */
+			break;
+		}
+
+		case TYPEFUNC_RECORD:
+		{
+			/* failed to determine actual type of RECORD */
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("function returning record called in context "
+							"that cannot accept type record")));
+			break;
+		}
+
+		default:
+		{
+			/* result type isn't composite */
+			elog(ERROR, "return type must be a row type");
+			break;
+		}
+	}
+	return returnSetInfo;
+}
+
+
+/*
+ * SetupTuplestore sets up a tuplestore for returning data.
+ */
+Tuplestorestate *
+SetupTuplestore(FunctionCallInfo fcinfo, TupleDesc *tupleDescriptor)
+{
+	MemoryContext perQueryContext = NULL;
+	MemoryContext oldContext = NULL;
+	Tuplestorestate *tupstore = NULL;
+	ReturnSetInfo *resultSet = CheckTuplestoreReturn(fcinfo, tupleDescriptor);
+
+	/* NOTE: Is this copy actually needed? */
+	*tupleDescriptor = CreateTupleDescCopy(*tupleDescriptor);
```

Comment:
This was done here: https://github.com/citusdata/citus/pull/2898/files#diff-77e7988c2a4f80c856387d90e08229cfL773

So I coppied it, but in all the other places it's not done. Is there a reason for this? It seems not needed to me.

Reply:
seems unnecessary

Reply:
I'll remove it then.